text/0037-guarded-arith.md

```diff
@@ -0,0 +1,82 @@
+# Guard Arithmetic Expressions within Policies
+
+## Related issues and PRs
+
+- Reference Issues: No existing issue, but supersedes https://github.com/cedar-policy/rfcs/pull/36
+- Implementation PR(s): (Existing PR for range analysis based on schema: https://github.com/cedar-policy/cedar/pull/56)
+
+## Timeline
+
+- Started: 2023-11-07
+
+## Summary
+
+We will use a range analysis (as in https://github.com/cedar-policy/cedar/pull/56) but with the bounds given in the policy rather than in the schema.
+
+## Motivation
+
+Arithmetic with bounded integers is a common source of bugs [https://blog.research.google/2006/06/extra-extra-read-all-about-it-nearly.html]. Consider the following policy:
+
+```
+forbid(principal, action, resource) when {
+   principal.elevated_privileges_end_time_s * 1000 > context.current_time_ms ||
+   !principal.is_mfa
+};
+```
+
+A casual reader of this policy may conclude that all actions are forbiden when `principal.mfa` is `false`. This is wrong. The lhs (`principal.elevated_privileges_end_time_s * 1000 > context.current_time_ms`) is evaluated first and if it errors the entire policy is skipped.
+
+To avoid this class of mistakes, we want to change Cedar's validator to error on arithmetic expressions that could overflow. To pass validation, arithmetic should be guarded. For example, the above policy should be written:
+
+```
+forbid(principal, action, resource) when {
+   (principal.elevated_privileges_end_time_s < i64::MAX / 1000 &&
+    principal.elevated_privileges_end_time_s > i64::MIN / 1000 &&
```

Comment:
Are you indirectly proposing to add `i64::MAX` and `i64::MIN` to the language? Without these things, writing inline bounds checks will be a hassle. Assuming, therefore, you _are_ proposing to add them, I suggest we not use namespace syntax, which implies `i64::MAX` is a type, not a value. Want to think of some proposals for alternative syntax?

Reply:
I don't think we should add these constants to the language.  They aren't useful for writing range constraints in that they are too lenient to allow any non-trivial arithmetic to remain within bounds.

Plus, we definitely don't want to allow division :)

Reply:
Yeah, I just wrote this to make the example easier to read, no intent to add to the language